stdlib/dynarray.ml

```diff
@@ -0,0 +1,758 @@
+     the loop (before or after, or both). However, notice that this is
+     not necessary to detect the removal of elements from [a]: if
+     elements have been removed by the time the [for] loop reaches
+     them, then [unsafe_get] will itself fail with an [Invalid_argument]
+     exception. We only need to detect the addition of new elements to
+     [a] during iteration, and for this it is enough to call
+     [check_same_length] once at the end.
+
+     Calling [check_same_length] more often could catch more
+     programming errors, but the only errors that we miss with this
+     optimization are those that keep the array size constant --
+     additions and deletions that cancel each other. We consider this
+     an acceptable tradeoff.
+  *)
+  check_valid_length length arr;
+  for i = 0 to length - 1 do
+    k (unsafe_get arr ~i ~length);
+  done;
+  check_same_length f a ~length
+
+let iter k a =
+  iter_ "iter" k a
+
+let iteri k a =
+  let {arr; length} = a in
+  check_valid_length length arr;
+  for i = 0 to length - 1 do
+    k i (unsafe_get arr ~i ~length);
+  done;
+  check_same_length "iteri" a ~length
+
+let map f a =
+  let {arr; length} = a in
+  check_valid_length length arr;
+  let res = {
+    length;
+    arr = Array.init length (fun i ->
+      Elem {v = f (unsafe_get arr ~i ~length)});
+  } in
+  check_same_length "map" a ~length;
+  res
+
+
+let mapi f a =
+  let {arr; length} = a in
+  check_valid_length length arr;
+  let res = {
+    length;
+    arr = Array.init length (fun i ->
+      Elem {v = f i (unsafe_get arr ~i ~length)});
+  } in
+  check_same_length "mapi" a ~length;
+  res
+
+let fold_left f acc a =
+  let {arr; length} = a in
+  check_valid_length length arr;
+  let r = ref acc in
+  for i = 0 to length - 1 do
+    let v = unsafe_get arr ~i ~length in
+    r := f !r v;
+  done;
+  check_same_length "fold_left" a ~length;
+  !r
+
+let fold_right f a acc =
+  let {arr; length} = a in
+  check_valid_length length arr;
+  let r = ref acc in
+  for i = length - 1 downto 0 do
+    let v = unsafe_get arr ~i ~length in
+    r := f v !r;
+  done;
+  check_same_length "fold_right" a ~length;
+  !r
+
+let exists p a =
+  let {arr; length} = a in
+  check_valid_length length arr;
+  let rec loop p arr i length =
+    if i = length then false
+    else
+      p (unsafe_get arr ~i ~length)
+      || loop p arr (i + 1) length
+  in
+  let res = loop p arr 0 length in
+  check_same_length "exists" a ~length;
+  res
+
+let for_all p a =
+  let {arr; length} = a in
+  check_valid_length length arr;
+  let rec loop p arr i length =
+    if i = length then true
+    else
+      p (unsafe_get arr ~i ~length)
+      && loop p arr (i + 1) length
+  in
+  let res = loop p arr 0 length in
+  check_same_length "for_all" a ~length;
+  res
+
+let filter f a =
+  let b = create () in
+  iter_ "filter" (fun x -> if f x then add_last b x) a;
+  b
+
+let filter_map f a =
+  let b = create() in
```

Comment:
Very minor, but usually there's a space in `create ()`, except for here and in `of_seq`.